core/arns/src/main/java/software/amazon/awssdk/arns/Arn.java

```diff
@@ -0,0 +1,372 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.arns;
+
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * The Arns generated and recognized by this code are the Arns described here:
+ *
+ * https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+ *
+ * <p>
+ * The primary supported Arn format is:
+ *
+ * <code>
+ * arn::&#60;partition&#62;::&#60;service&#62;::&#60;region&#62;::&#60;account&#62;::&#60;resource&#62;
+ * </code>
+ *
+ * <p>
+ * {@link #resourceAsString()} returns everything after the account section of the Arn
+ * as a single string.
+ *
+ * <p>
+ * However, the following Arn formats are supported where the values are present and well
+ * formatted through {@link #resource()}:
+ *
+ * <pre>
+ * arn:&#60;partition&#62;:&#60;service&#62;:&#60;region&#62;:&#60;account&#62;:&#60;resourcetype&#62;/resource
+ * arn:&#60;partition&#62;:&#60;service&#62;:&#60;region&#62;:&#60;account&#62;:&#60;resourcetype&#62;/resource/qualifier
+ * arn:&#60;partition&#62;:&#60;service&#62;:&#60;region&#62;:&#60;account&#62;:&#60;resourcetype&#62;/resource:qualifier
+ * arn:&#60;partition&#62;:&#60;service&#62;:&#60;region&#62;:&#60;account&#62;:&#60;resourcetype&#62;:resource
+ * arn:&#60;partition&#62;:&#60;service&#62;:&#60;region&#62;:&#60;account&#62;:&#60;resourcetype&#62;:resource:qualifier
+ * </pre>
+ *
+ * {@link #resource()} returns a {@link ArnResource} which has access
+ * to {@link ArnResource#resourceType()}, {@link ArnResource#resource()} and
+ * {@link ArnResource#qualifier()}.
+ *
+ * <p>
+ * To parse an Arn from a string use Arn.fromString(). To convert an Arn to it's
+ * string representation use Arn.toString().
+ *
+ * <p>
+ * For instance, for a string s, containing a well-formed Arn the
+ * following should always be true:
+ *
+ * <pre>
+ * Arn theArn = Arn.fromString(s);
+ * s.equals(theArn.toString());
+ * </pre>
+ *
+ * @see ArnResource
+ */
+@SdkPublicApi
+public final class Arn implements ToCopyableBuilder<Arn.Builder, Arn> {
+
+    private final String partition;
+    private final String service;
+    private final String region;
+    private final String accountId;
+    private final String resource;
+    private final ArnResource arnResource;
+
+    private Arn(DefaultBuilder builder) {
+        this.partition = Validate.paramNotBlank(builder.partition, "partition");
+        this.service = Validate.paramNotBlank(builder.service, "service");
+        this.region = Validate.paramNotBlank(builder.region, "region");
+        this.accountId = Validate.paramNotBlank(builder.accountId, "accountId");
```

Comment:
Why did we change the contract from the v1 Arn here where these are optional fields?

Reply:
This is my fault. Updated.